examples/contrib/oed/eig_estimation_benchmarking.py

```diff
@@ -7,148 +7,383 @@
-
-
-def naive_rainforth_lm(X_lm, w_sds, N, M):
-    model, _ = zero_mean_unit_obs_sd_lm(w_sds)
-    return naive_rainforth(model, X_lm, "y", "w", N=N, M=M)
-
-
-def donsker_varadhan_lm(X, w_sds, n_iter, n_samples, lr, T,
-                        final_X=None, final_n_samples=None, return_history=False):
-    model, _ = zero_mean_unit_obs_sd_lm(w_sds)
-    if final_X is None:
-        final_X = X
-    if final_n_samples is None:
-        final_n_samples = n_samples
-    dv_loss_fn = donsker_varadhan_loss(model, "y", T)
-    params = None
-    opt = optim.ClippedAdam({"lr": lr, "betas": (0.92, 0.999)})
-    history = []
-    for step in range(n_iter):
-        if params is not None:
-            pyro.infer.util.zero_grads(params)
-        agg_loss, dv_loss = dv_loss_fn(X, n_samples)
-        agg_loss.backward()
-        if return_history:
-            history.append(dv_loss)
-        params = [pyro.param(name).unconstrained()
-                  for name in pyro.get_param_store().get_all_param_names()]
-        opt(params)
-    _, dv_loss = dv_loss_fn(final_X, final_n_samples)
-    if return_history:
-        return torch.stack(history), dv_loss
+elbo = TraceEnum_ELBO(strict_enumeration_warning=False).differentiable_loss
+
+
+def linear_model_ground_truth(model, design, observation_labels, target_labels, eig=True):
+    if isinstance(target_labels, str):
+        target_labels = [target_labels]
+
+    w_sd = torch.cat(list(model.w_sds.values()), dim=-1)
+    prior_cov = torch.diag(w_sd**2)
+    posterior_covs = [analytic_posterior_cov(prior_cov, x, model.obs_sd) for x in torch.unbind(design)]
+    target_indices = get_indices(target_labels, tensors=model.w_sds)
+    target_posterior_covs = [S[target_indices, :][:, target_indices] for S in posterior_covs]
+    if eig:
+        prior_entropy = lm_H_prior(model, design, observation_labels, target_labels)
+        return prior_entropy - torch.tensor([0.5*torch.logdet(2*np.pi*np.e*C) for C in target_posterior_covs])
+    else:
+        return torch.tensor([0.5*torch.logdet(2*np.pi*np.e*C) for C in target_posterior_covs])
+
+
+def lm_H_prior(model, design, observation_labels, target_labels):
+    if isinstance(target_labels, str):
+        target_labels = [target_labels]
+
+    w_sd = torch.cat(list(model.w_sds.values()), dim=-1)
+    prior_cov = torch.diag(w_sd**2)
+    target_indices = get_indices(target_labels, tensors=model.w_sds)
+    target_prior_covs = prior_cov[target_indices, :][:, target_indices]
+    return 0.5*torch.logdet(2*np.pi*np.e*target_prior_covs)
+
+
+def mc_H_prior(model, design, observation_labels, target_labels, num_samples=1000):
+    if isinstance(target_labels, str):
+        target_labels = [target_labels]
+
+    expanded_design = lexpand(design, num_samples)
+    trace = pyro.poutine.trace(model).get_trace(expanded_design)
+    trace.compute_log_prob()
+    lp = sum(trace.nodes[l]["log_prob"] for l in target_labels)
+    return -lp.sum(0)/num_samples
```

Comment:
These prior entropy estimators are used to switch between APE and EIG for plotting the final graphs. One uses lm formula, the other uses Monte Carlo